kura/org.eclipse.kura.asset.cloudlet.provider/src/main/java/org/eclipse/kura/internal/asset/cloudlet/AssetCloudlet.java

```diff
@@ -194,272 +117,196 @@ protected synchronized void deactivate(final ComponentContext componentContext)
-     * @param channelName
-     *            the channel name (might contain {@code _} for multiple reads)
-     * @throws NullPointerException
-     *             if any of the arguments is null
-     */
-    private void readChannelsByNames(final KuraResponsePayload respPayload, final String assetPid,
-            final String channelName) {
-        requireNonNull(respPayload, message.respPayloadNonNull());
-        requireNonNull(assetPid, message.assetPidNonNull());
-        requireNonNull(channelName, message.channelNameNonNull());
-
-        final String channelDelim = "_";
-        Set<String> channelNames = null;
-        if (channelName.contains(channelDelim)) {
-            channelNames = CollectionUtil.newHashSet(Arrays.asList(channelName.split(channelDelim)));
-            channelNames.removeAll(Collections.singleton(""));
+        JsonArray request;
+
+        try {
+            request = getRequestJsonArray(reqPayload);
+        } catch (Exception e) {
+            respPayload.setResponseCode(KuraResponsePayload.RESPONSE_CODE_BAD_REQUEST);
+            return;
         }
-        final Asset asset = this.assets.get(assetPid);
-        final AssetConfiguration configuration = asset.getAssetConfiguration();
-        final Map<String, Channel> assetConfiguredChannels = configuration.getAssetChannels();
 
-        final Set<String> channelNamesToRead = CollectionUtil.newHashSet();
-        if (channelNames == null) {
-            channelNamesToRead.add(channelName);
+        ChannelOperationResponse response;
+
+        if (request == null || request.size() == 0) {
+            response = readAllAssets(respPayload);
         } else {
-            for (final String chName : channelNames) {
-                channelNamesToRead.add(chName);
-            }
-        }
-        if (assetConfiguredChannels != null) {
-            List<ChannelRecord> channelRecords = null;
+            List<ReadRequest> readRequests;
             try {
-                channelRecords = asset.read(channelNamesToRead);
-            } catch (final KuraException e) {
-                // if connection exception occurs
-                logger.warn(message.connectionException() + e);
-                respPayload.addMetric(message.errorMessage(), message.connectionException());
+                readRequests = ReadRequest.parseAll(request);
+            } catch (Exception e) {
+                respPayload.setResponseCode(KuraResponsePayload.RESPONSE_CODE_BAD_REQUEST);
+                return;
             }
-            if (channelRecords != null) {
-                this.prepareResponse(respPayload, channelRecords);
+            response = new ChannelOperationResponse();
+            for (ReadRequest readRequest : readRequests) {
+                readAsset(readRequest.getAssetName(), readRequest.getChannelNames(), response);
             }
         }
+
+        respPayload.setBody(response.getSerialized().toString().getBytes(StandardCharsets.UTF_8));
```

Comment:
Maybe this `response.getSerialized().toString().getBytes(StandardCharsets.UTF_8)` can be done in a specific method in `ChannelOperationResponse `